rfcs/013-gitlab-spam-checker/013-gitlab-spam-checker.md

```diff
@@ -0,0 +1,93 @@
+
+Ideally, the solution will also be able to check abuse reports manually submitted by users in GitLab and check the reported URL in the abuse report for spam content.
+
+A feedback mechanism would also be implemented, allowing the machine learning model to learn from mistakes and improve over time. This could be done by storing content in an S3 bucket and regularly re-training the model on a dataset including new data. The system should also keep a log of decisions made and the confidence level of the decision.
+
+Similarly to GitLab's Spamcheck service, storing the model in an S3 bucket and pulling it to the spam management service on start-up would allow the model to be updated without having to redeploy the spam management service.
+
+An allowlist of trusted email domain providers could be implemented to prevent false positives to ensure users with trusted email domains can post content without it being checked for spam. YAT already possesses a list of trusted email domains.
+
+This mermaid diagram provides an overview of how this system would work:
+
+```mermaid
+  graph TB;
+    GH["GitLab Hooks"] -- "Triggers on Create/Update" --> SC["Spam Checking Service"]
+    AR["Abuse Reports"] -- "Retrieved from GitLab" --> SC
+    SC -- "Checks" --> I["Issues"]
+    SC -- "Checks" --> N["Notes"]
+    SC -- "Checks" --> P["Projects"]
+    SC -- "Checks" --> U["Users"]
+    S3ML["ML model stored in S3 Bucket"] -- "On Start-Up" --> SC
+    TED["Trusted Email Domains List"] -- "Skips Spam Checks" --> SC
+    SC -- "Invokes" --> ML["Machine Learning Model"]
+    ML -- "Classifies" --> SP["Spam"]
+    SP -- "Identified" --> SC
+    SC -- "Deletes via API" --> GD["GitLab"]
+    SJ["Snippet Job"] -. "Runs Periodically" .-> SC
+    SC -- "Logs Decisions & Confidence Levels" --> L["Logs"]
+    SC -- "Stores Evaluated Content" --> S3["S3 Bucket"]
+    linkStyle 0 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 1 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 2 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 3 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 4 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 5 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 6 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 7 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 8 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 9 stroke:#2ecd71,stroke-width:2px,dasharray: 5, 5;
+    linkStyle 10 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 11 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 12 stroke:#2ecd71,stroke-width:2px;
+```
+
+## Drawbacks
+
+1. Ensuring the system can handle many system hooks from GitLab without significant latency is critical. It needs to be scalable and able to handle peak loads.
```

Comment:
Meaning it should not impact the performance of gitlab, right? I think this should be made more explicit...

Reply:
For example, 80 issues can sometimes be added in a matter of seconds. We need to make sure the spam management service is available to accept that number of hooks in that time frame, otherwise, some might fall between the gaps

Reply:
It could be better but its abstract for a reason atm, I will evolve it as the design of the system becomes more clear

Reply:
oh, this is a good point to expand on the design: since we rely on events "push" (e.g. webhook incoming from GL) we need to make sure we store those in a stack and then process them pulling one by one. 
AFAIK an event based queue works well for this purpose, and the queue itself should be persisted (so if the service restarts it's not lost).

What Sean mentions, it's a different topic, but one that needs to be made explicit indeed: when the service acts on GL we should make sure that we won't hammer GL too much.

Reply:
Hopefully less of an issue with GLCN but yes indeed